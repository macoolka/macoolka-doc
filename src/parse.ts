/**
 * @file
 */
import { parse as _parse } from 'macoolka-type-ast';
import { Module } from 'macoolka-type-model';
import * as array from 'fp-ts/lib/Array';
import { pipe } from 'fp-ts/lib/pipeable';
import * as E from 'fp-ts/lib/Either';
import * as R from 'fp-ts/lib/Reader';
import { FileWhereUniqueInput } from 'macoolka-store-core';
const filterModule = (a: Module.MModule) => {

    const result: Module.MModule = {
        ...a,
        constants: pipe(
            a.constants,
            array.filter(a => a.isExported && a.ignore !== true)
        ),
        functions: pipe(
            a.functions,
            array.filter(a => a.isExported && a.ignore !== true)
        ),
        classes: pipe(
            a.classes,
            array.filter(a => a.isExported && a.ignore !== true)
        ),
        interfaces: pipe(
            a.interfaces,
            array.filter(a => a.isExported && a.ignore !== true)
        ),
        typealiases: pipe(
            a.typealiases,
            array.filter(a => a.isExported && a.ignore !== true)
        ),
        exports: pipe(
            a.exports,
            array.filter(a => a.ignore !== true)
        ),
    };
    return result;
};
/**
 * Parse files to Module
 * @desczh
 * 解析文件到Module
 * @since 0.2.0
 */
export const parse = (files: FileWhereUniqueInput[]) =>
    pipe(
        _parse(files),
        E.map(as =>
            pipe(
                as,
                R.map(a =>
                    pipe(
                        a,
                        array.filter(a => a.file === true),
                        array.map(filterModule)
                    ))
            ))
    );
