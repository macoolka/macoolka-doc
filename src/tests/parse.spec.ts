import { parse } from '../parse'
import { writeFileSync } from 'fs'
import {pipe} from 'fp-ts/lib/pipeable'
import * as E from 'fp-ts/lib/Either'
//import {formatWithType} from 'macoolka-format'
describe('typescript get tags', () => {
    it('parse default', () => {
     
        pipe(
            parse([{ container:__dirname, folders:['fixtures'],name:'Basic.ts' }]),
            E.map(a=>{
                const value=a({})
                writeFileSync(__dirname + '/Basic.json', JSON.stringify(value), 'utf-8')
                expect(value).toMatchSnapshot();
                return a
            }),
            a=>expect(E.isRight(a)).toBeTruthy()
        )
    })
    it('parse locale zh', () => {
  
        pipe(
            parse([{ container:__dirname, folders:['fixtures'],name:'Basic.ts' }]),
            E.map(a=>{
                const value=a({i18n:{locale:'zh'}})
                writeFileSync(__dirname + '/Basic.zh.json', JSON.stringify(value), 'utf-8')
                expect(value).toMatchSnapshot();
                return a
            }),
            a=>expect(E.isRight(a)).toBeTruthy()
        )
     
    })

})