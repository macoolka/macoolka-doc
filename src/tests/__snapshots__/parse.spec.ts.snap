// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typescript get tags parse default 1`] = `
Array [
  Object {
    "_kind": "module",
    "classes": Array [
      Object {
        "_kind": "class",
        "deprecated": false,
        "description": Array [],
        "examples": Array [],
        "fields": Array [
          Object {
            "_kind": "field",
            "deprecated": false,
            "description": Array [
              "class field desc",
            ],
            "examples": Array [],
            "exclusiveCreate": false,
            "exclusiveLoad": false,
            "exclusiveSearch": false,
            "exclusiveUpdate": false,
            "exclusiveWhere": false,
            "id": false,
            "ignore": false,
            "name": "classfield",
            "order": true,
            "path": Array [],
            "readonly": false,
            "reason": Array [],
            "required": false,
            "signature": "classfield: string = ''",
            "since": "0.2.0",
            "type": "string",
            "unique": false,
          },
        ],
        "ignore": false,
        "implements": Array [],
        "isExported": true,
        "methods": Array [
          Object {
            "_kind": "method",
            "deprecated": false,
            "description": Array [
              "class method desc",
            ],
            "examples": Array [],
            "ignore": false,
            "name": "classMethod",
            "params": Array [],
            "path": Array [],
            "reason": Array [],
            "required": false,
            "returnVoid": false,
            "signature": "classMethod( { ... }",
            "since": "0.2.0",
            "type": "string",
          },
        ],
        "name": "Class",
        "path": Array [],
        "reason": Array [],
        "signature": "export class Class { ... }",
        "since": "0.2.0",
        "staticMethods": Array [
          Object {
            "_kind": "staticmethod",
            "deprecated": false,
            "description": Array [
              "class static method desc",
            ],
            "examples": Array [],
            "ignore": false,
            "name": "classStaticMethod",
            "params": Array [],
            "path": Array [],
            "reason": Array [],
            "required": false,
            "returnVoid": false,
            "signature": "static classStaticMethod( { ... }",
            "since": "0.2.0",
            "type": "string",
          },
        ],
      },
    ],
    "constants": Array [
      Object {
        "_kind": "constant",
        "deprecated": false,
        "description": Array [
          "constant desc",
        ],
        "examples": Array [],
        "ignore": false,
        "isExported": true,
        "name": "constanta",
        "path": Array [],
        "reason": Array [],
        "required": false,
        "signature": "export const constanta: 1 = ...",
        "since": "0.2.0",
        "type": "string",
      },
      Object {
        "_kind": "constant",
        "deprecated": false,
        "description": Array [
          "inner constant desc",
        ],
        "examples": Array [],
        "ignore": false,
        "isExported": true,
        "name": "exportvara",
        "path": Array [],
        "reason": Array [],
        "required": false,
        "signature": "export const exportvara: \\"\\" = ...",
        "since": "0.2.0",
        "type": "string",
      },
    ],
    "deprecated": false,
    "description": Array [],
    "examples": Array [],
    "exports": Array [
      Object {
        "_kind": "export",
        "deprecated": false,
        "description": Array [
          "export function desc",
        ],
        "examples": Array [],
        "ignore": false,
        "name": "exportmethod",
        "path": Array [],
        "reason": Array [],
        "signature": "() => any",
        "since": "0.2.0",
      },
      Object {
        "_kind": "export",
        "deprecated": false,
        "description": Array [
          "export constant desc",
        ],
        "examples": Array [],
        "ignore": false,
        "name": "exportvara",
        "path": Array [],
        "reason": Array [],
        "signature": "\\"\\"",
        "since": "0.2.0",
      },
    ],
    "file": true,
    "functions": Array [
      Object {
        "_kind": "function",
        "deprecated": false,
        "description": Array [
          "export inner function desc",
        ],
        "examples": Array [],
        "ignore": false,
        "isExported": true,
        "name": "exportmethod",
        "params": Array [],
        "path": Array [],
        "reason": Array [],
        "required": false,
        "returnVoid": false,
        "signature": "export const exportmethod = () => ...",
        "since": "0.2.0",
        "type": "string",
      },
      Object {
        "_kind": "function",
        "deprecated": false,
        "description": Array [
          "function desc",
        ],
        "examples": Array [],
        "ignore": false,
        "isExported": true,
        "name": "functiona",
        "params": Array [],
        "path": Array [],
        "reason": Array [],
        "required": false,
        "returnVoid": false,
        "signatures": Array [
          "export function functiona( { ... }",
        ],
        "since": "0.2.0",
        "type": "string",
      },
    ],
    "idUnique": false,
    "ignore": false,
    "interfaces": Array [],
    "name": "Basic",
    "path": Array [
      "",
      "home",
      "fastspeeed",
      "macoolka",
      "macoolka-doc",
      "src",
      "tests",
      "fixtures",
      "Basic.ts",
    ],
    "reason": Array [],
    "since": "0.2.0",
    "typealiases": Array [
      Object {
        "_kind": "typealias",
        "deprecated": false,
        "description": Array [
          "typeAlias desc",
        ],
        "examples": Array [],
        "fields": Array [],
        "ignore": false,
        "isExported": true,
        "methods": Array [],
        "name": "Type",
        "path": Array [],
        "reason": Array [],
        "required": false,
        "signature": "export type Type = {


}",
        "since": "0.2.0",
        "type": "string",
      },
    ],
  },
]
`;

exports[`typescript get tags parse locale zh 1`] = `
Array [
  Object {
    "_kind": "module",
    "classes": Array [
      Object {
        "_kind": "class",
        "deprecated": false,
        "description": Array [],
        "examples": Array [],
        "fields": Array [
          Object {
            "_kind": "field",
            "deprecated": false,
            "description": Array [
              "类字段描述",
            ],
            "examples": Array [],
            "exclusiveCreate": false,
            "exclusiveLoad": false,
            "exclusiveSearch": false,
            "exclusiveUpdate": false,
            "exclusiveWhere": false,
            "id": false,
            "ignore": false,
            "name": "classfield",
            "order": true,
            "path": Array [],
            "readonly": false,
            "reason": Array [],
            "required": false,
            "signature": "classfield: string = ''",
            "since": "0.2.0",
            "type": "string",
            "unique": false,
          },
        ],
        "ignore": false,
        "implements": Array [],
        "isExported": true,
        "methods": Array [
          Object {
            "_kind": "method",
            "deprecated": false,
            "description": Array [
              "类方法描述",
            ],
            "examples": Array [],
            "ignore": false,
            "name": "classMethod",
            "params": Array [],
            "path": Array [],
            "reason": Array [],
            "required": false,
            "returnVoid": false,
            "signature": "classMethod( { ... }",
            "since": "0.2.0",
            "type": "string",
          },
        ],
        "name": "Class",
        "path": Array [],
        "reason": Array [],
        "signature": "export class Class { ... }",
        "since": "0.2.0",
        "staticMethods": Array [
          Object {
            "_kind": "staticmethod",
            "deprecated": false,
            "description": Array [
              "类静态方法描述",
            ],
            "examples": Array [],
            "ignore": false,
            "name": "classStaticMethod",
            "params": Array [],
            "path": Array [],
            "reason": Array [],
            "required": false,
            "returnVoid": false,
            "signature": "static classStaticMethod( { ... }",
            "since": "0.2.0",
            "type": "string",
          },
        ],
      },
    ],
    "constants": Array [
      Object {
        "_kind": "constant",
        "deprecated": false,
        "description": Array [
          "变量描述",
        ],
        "examples": Array [],
        "ignore": false,
        "isExported": true,
        "name": "constanta",
        "path": Array [],
        "reason": Array [],
        "required": false,
        "signature": "export const constanta: 1 = ...",
        "since": "0.2.0",
        "type": "string",
      },
      Object {
        "_kind": "constant",
        "deprecated": false,
        "description": Array [
          "变量描述",
        ],
        "examples": Array [],
        "ignore": false,
        "isExported": true,
        "name": "exportvara",
        "path": Array [],
        "reason": Array [],
        "required": false,
        "signature": "export const exportvara: \\"\\" = ...",
        "since": "0.2.0",
        "type": "string",
      },
    ],
    "deprecated": false,
    "description": Array [
      "模块描述",
    ],
    "examples": Array [],
    "exports": Array [
      Object {
        "_kind": "export",
        "deprecated": false,
        "description": Array [
          "导出函数描述",
        ],
        "examples": Array [],
        "ignore": false,
        "name": "exportmethod",
        "path": Array [],
        "reason": Array [],
        "signature": "() => any",
        "since": "0.2.0",
      },
      Object {
        "_kind": "export",
        "deprecated": false,
        "description": Array [
          "导出变量描述",
        ],
        "examples": Array [],
        "ignore": false,
        "name": "exportvara",
        "path": Array [],
        "reason": Array [],
        "signature": "\\"\\"",
        "since": "0.2.0",
      },
    ],
    "file": true,
    "functions": Array [
      Object {
        "_kind": "function",
        "deprecated": false,
        "description": Array [
          "函数描述",
        ],
        "examples": Array [],
        "ignore": false,
        "isExported": true,
        "name": "exportmethod",
        "params": Array [],
        "path": Array [],
        "reason": Array [],
        "required": false,
        "returnVoid": false,
        "signature": "export const exportmethod = () => ...",
        "since": "0.2.0",
        "type": "string",
      },
      Object {
        "_kind": "function",
        "deprecated": false,
        "description": Array [
          "函数描述",
        ],
        "examples": Array [],
        "ignore": false,
        "isExported": true,
        "name": "functiona",
        "params": Array [],
        "path": Array [],
        "reason": Array [],
        "required": false,
        "returnVoid": false,
        "signatures": Array [
          "export function functiona( { ... }",
        ],
        "since": "0.2.0",
        "type": "string",
      },
    ],
    "idUnique": false,
    "ignore": false,
    "interfaces": Array [],
    "name": "Basic",
    "path": Array [
      "",
      "home",
      "fastspeeed",
      "macoolka",
      "macoolka-doc",
      "src",
      "tests",
      "fixtures",
      "Basic.ts",
    ],
    "reason": Array [],
    "since": "0.2.0",
    "typealiases": Array [
      Object {
        "_kind": "typealias",
        "deprecated": false,
        "description": Array [
          "类型描述",
        ],
        "examples": Array [],
        "fields": Array [],
        "ignore": false,
        "isExported": true,
        "methods": Array [],
        "name": "Type",
        "path": Array [],
        "reason": Array [],
        "required": false,
        "signature": "export type Type = {


}",
        "since": "0.2.0",
        "type": "string",
      },
    ],
  },
]
`;
